src/d3d8/d3d8_device.cpp

```diff
@@ -1680,16 +1696,29 @@ namespace dxvk {
     if (unlikely(pDeclaration == nullptr || pHandle == nullptr))
       return D3DERR_INVALIDCALL;
 
+    // Validate VS version for non-FF shaders
+    if (pFunction != nullptr) {
+      uint32_t majorVersion = (pFunction[0] >> 8) & 0xff;
+      uint32_t minorVersion = pFunction[0] & 0xff;
+
+      if (unlikely(majorVersion != 1 || minorVersion > 1)) {
+        Logger::err(str::format("D3D8Device::CreateVertexShader: Unsupported VS version ", majorVersion, ".", minorVersion));
+        return D3DERR_INVALIDCALL;
+      }
+    }
+
     D3D8VertexShaderInfo& info = m_vertexShaders.emplace_back();
 
     // Store D3D8 bytecodes in the shader info
-    if (pDeclaration != nullptr)
-      for (UINT i = 0; pDeclaration[i+1] != D3DVSD_END(); i++)
-        info.declaration.push_back(pDeclaration[i]);
+    for (UINT i = 0; pDeclaration[i] != D3DVSD_END(); i++)
+      info.declaration.push_back(pDeclaration[i]);
+    info.declaration.push_back(D3DVSD_END());
```

Comment:
The old code was straight up broken, right?

`      for (UINT i = 0; pDeclaration[i+1] != D3DVSD_END(); i++)`

This misses the last token **before** the end and the end token. The new code makes sense to me, I'd just like to make sure there isn't some clever intention behind the old code that you accidentally removed here.

Reply:
Yep, was pretty much broken since the entire definition/function needed to be stored, including the end token. Nothing seems to care about this, but luckily there were a few wine tests that helped me spot it.

Reply:
Interesting, I guess if any game actually used the declaration from GetVertexVertexShader it would've read way out of bounds.